docs/refs/api/rtos/callbacks.md

```diff
@@ -1,9 +1,384 @@
-#### Callbacks
+### Callbacks
 
-[Document about callbacks, their best practices and dos and don'ts.]
+A callback is a user-provided piece of code that is passed to another function. The callback allows an API to execute the user’s code in its own context.
 
-###### Thread example with callbacks
+```c++
+// Create a serial object
+Serial serial(USBTX, USBRX);
 
-The Callback API provides a convenient way to pass arguments to spawned threads.  
+// A function that echoes any received data back
+ void echo() {
+    while (serial.readable()) {
+         serial.putc(serial.getc());
+    }
+ }
 
-[![View code](https://www.mbed.com/embed/?url=https://developer.mbed.org/teams/mbed_example/code/rtos_threading_with_callback/)](https://developer.mbed.org/teams/mbed_example/code/rtos_threading_with_callback/file/d4b2a035ffe3/main.cpp)
+ int main(void) {
+     // Call our function echo whenever the serial line receives data
+     serial.attach(&echo, Serial::RxIrq);
+ }
+```
+
+#### The importance of state
+
+Callbacks have two important pieces of information, the code to execute and the state associated with the callback.
+
+Without state, callbacks offer limited composability. In C, you can fix this by adding an additional “state” argument to the function pointer, so you can pass in the opaque “state” when you register the callback.
+
```

Comment:
Why remove the ADC examples?
1. Theres no longer an example of the limitations that come with not supporting state. Most of the value of the Callback class is lost if this point is missed. 
2. We are not covering all of the function types that the Callback class supports.

If we're moving these elsewhere we should probably just remove these sections (importance of state and separate callback class), since they only have limited value without showing all of the options available for callbacks.

We've also removed the motivation for introducing the Callback class. Is this going elsewhere?